File: easy/src/fractal/Variation.scala
package fractal

import scala.math._

object Variation{
    def v0(p: Point) = p
    def v1(p: Point) = new Point(sin(p.x), sin(p.y))
    def v2(p: Point) = p / p.r2()
    def v3(p: Point) = {
        val (x, y) = (p.x, p.y)
        val r2 = p.r2()
        new Point(x*sin(r2)-y*cos(r2), x*cos(r2)+y*sin(r2))
    }
    def v4(p: Point) = {
        val (x, y) = (p.x, p.y)
        new Point((x-y)*(x+y), 2*x*y) / p.r()
    }
    def v5(p: Point) = new Point(p.theta() / Pi, p.r() - 1)
    def v6(p: Point) = {
        val (r, theta) = (p.r(), p.theta())
        new Point(sin(theta+r), cos(theta-r)) * r
    }
    def v7(p: Point) = {
        val (r, theta) = (p.r(), p.theta())
        new Point(sin(theta*r), -cos(theta*r)) * r
    }
    def v8(p: Point) = {
        val r = p.r()
        new Point(sin(Pi*r), cos(Pi*r)) * (p.theta()/Pi)
    }
    def v9(p: Point) = {
        val (r, theta) = (p.r(), p.theta())
        new Point(cos(theta)+sin(r), sin(theta)-cos(r)) / r
    }
    def v10(p: Point) = {
        val (r, theta) = (p.r(), p.theta())
        new Point(sin(theta)/r, r*cos(theta))
    }
    def v11(p: Point) = {
        val (r, theta) = (p.r(), p.theta())
        new Point(sin(theta)*cos(r), cos(theta)*sin(r))
    }
    def v12(p: Point) = {
        val (r, theta) = (p.r(), p.theta())
        val p0 = sin(theta+r)
        val p1 = cos(theta-r)
        val p03 = p0 * p0 * p0
        val p13 = p1 * p1 * p1
        new Point(p03+p13, p03-p13) * r
    }

    val variations: Array[Point => Point] = Array(v0,v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12)
}

File: IDE/js/Variation.js
const variations = (function(){
    var pi = 3.14159265358979323846;

    function v0(p){
        return p;
    }

    function v1(p){
        return p.apply(Math.sin);
    }

    function v2(p){
        return p.scale(1 / p.r2());
    }

    function v3(p){
        var r2 = p.r2();
        return new Point(p.x * Math.sin(r2) - p.y * Math.cos(r2), p.x * Math.cos(r2) + p.y * Math.sin(r2));
    }

    function v4(p){
        var r = p.r();
        return new Point((p.x - p.y) * (p.x + p.y) / r, 2 * p.x * p.y / r);
    }

    function v5(p){
        var theta = p.theta();
        var r = p.r();
        return new Point(theta / pi, r - 1);
    }

    function v6(p){
        var r = p.r();
        var theta = p.theta();
        return new Point(r * Math.sin(theta + r), r * Math.cos(theta - r));
    }

    function v7(p){
        var r = p.r();
        var theta = p.theta();
        return new Point(r * Math.sin(theta * r), r * (- Math.cos(theta * r)));
    }

    function v8(p){
        var r = p.r();
        var theta = p.theta();
        return (new Point(Math.sin(pi * r), Math.cos(pi * r))).scale(theta / pi);
    }

    function v9(p){
        var r = p.r();
        var theta = p.theta();
        return (new Point(Math.cos(theta) + Math.sin(r), Math.sin(theta) - Math.cos(r))).scale(1 / r);
    }

    function v10(p){
        var r = p.r();
        var theta = p.theta();
        return new Point(Math.sin(theta) / r, r * Math.cos(theta));
    }

    function v11(p){
        var r = p.r();
        var theta = p.theta();
        return new Point(Math.sin(theta) * Math.cos(r), Math.cos(theta) * Math.sin(r));
    }

    function v12(p){
        var r = p.r();
        var theta = p.theta();
        var p03 = Math.pow(Math.sin(theta + r), 3.0);
        var p13 = Math.pow(Math.cos(theta - r), 3.0);
        return new Point(r * (p03 + p13), r * (p03 - p13));
    }

    return [v0,v1,v2,v3,v4,v5,v6,v7,v8,v9,v10,v11,v12];

})();
